enterprise/dbcrypt/dbcrypt.go

```diff
@@ -0,0 +1,374 @@
+	link, err := db.Store.GetGitAuthLink(ctx, params)
+	if err != nil {
+		return database.GitAuthLink{}, err
+	}
+	if err := db.decryptField(&link.OAuthAccessToken, link.OAuthAccessTokenKeyID); err != nil {
+		return database.GitAuthLink{}, err
+	}
+	if err := db.decryptField(&link.OAuthRefreshToken, link.OAuthRefreshTokenKeyID); err != nil {
+		return database.GitAuthLink{}, err
+	}
+	return link, nil
+}
+
+func (db *dbCrypt) GetGitAuthLinksByUserID(ctx context.Context, userID uuid.UUID) ([]database.GitAuthLink, error) {
+	links, err := db.Store.GetGitAuthLinksByUserID(ctx, userID)
+	if err != nil {
+		return nil, err
+	}
+	for idx := range links {
+		if err := db.decryptField(&links[idx].OAuthAccessToken, links[idx].OAuthAccessTokenKeyID); err != nil {
+			return nil, err
+		}
+		if err := db.decryptField(&links[idx].OAuthRefreshToken, links[idx].OAuthRefreshTokenKeyID); err != nil {
+			return nil, err
+		}
+	}
+	return links, nil
+}
+
+func (db *dbCrypt) UpdateGitAuthLink(ctx context.Context, params database.UpdateGitAuthLinkParams) (database.GitAuthLink, error) {
+	if err := db.encryptField(&params.OAuthAccessToken, &params.OAuthAccessTokenKeyID); err != nil {
+		return database.GitAuthLink{}, err
+	}
+	if err := db.encryptField(&params.OAuthRefreshToken, &params.OAuthRefreshTokenKeyID); err != nil {
+		return database.GitAuthLink{}, err
+	}
+	link, err := db.Store.UpdateGitAuthLink(ctx, params)
+	if err != nil {
+		return database.GitAuthLink{}, err
+	}
+	if err := db.decryptField(&link.OAuthAccessToken, link.OAuthAccessTokenKeyID); err != nil {
+		return database.GitAuthLink{}, err
+	}
+	if err := db.decryptField(&link.OAuthRefreshToken, link.OAuthRefreshTokenKeyID); err != nil {
+		return database.GitAuthLink{}, err
+	}
+	return link, nil
+}
+
+func (db *dbCrypt) encryptField(field *string, digest *sql.NullString) error {
+	// If no cipher is loaded, then we can't encrypt anything!
+	if db.ciphers == nil || db.primaryCipherDigest == "" {
+		return ErrNotEnabled
+	}
+
+	if field == nil {
+		return nil
+	}
+
+	encrypted, err := db.ciphers[db.primaryCipherDigest].Encrypt([]byte(*field))
+	if err != nil {
+		return err
+	}
+	// Base64 is used to support UTF-8 encoding in PostgreSQL.
+	*field = b64encode(encrypted)
+	*digest = sql.NullString{String: db.primaryCipherDigest, Valid: true}
+	return nil
+}
+
+// decryptFields decrypts the given field using the key with the given digest.
+// If the value fails to decrypt, sql.ErrNoRows will be returned.
+func (db *dbCrypt) decryptField(field *string, digest sql.NullString) error {
+	if db.ciphers == nil {
+		return ErrNotEnabled
+	}
+
+	if !digest.Valid || digest.String == "" {
+		// This field is not encrypted.
+		return nil
+	}
+
+	if field == nil || len(*field) == 0 {
+		// We've been asked to decrypt a field that is empty.
+		// There is a digest present, so it should have been encrypted,
+		// which would have produced a non-empty value.
```

Comment:
Our current implementation adds the nonce to the ciphertext, so what we store in the database should always end up being non-empty after encryption. I would assume that any future encryption algorithms we use here would also need to have a nonce.

Reply:
It's likely they'll be some sort of authenticated encryption, but you kinda never know what customers are going to demand.

To me it feels like a layering violation to police "acceptable" ciphertexts here, instead of depending on the ciphers themselves.